File: C++/src/merge_utils.cpp
#include "merge_utils.hpp"

void MergeUtils::merge(std::vector<int> &vec, size_t start, size_t mid, size_t end)
{
  size_t n1 = mid - start + 1;
  size_t n2 = end - mid;

  std::vector<int> vec1(n1);
  std::vector<int> vec2(n2);

  for (size_t i = 0; i < n1; i++)
  {
    vec1[i] = vec[start + i];
  }

  for (size_t i = 0; i < n2; i++)
  {
    vec2[i] = vec[mid + 1 + i];
  }

  size_t i = 0;
  size_t j = 0;
  size_t k = start;

  while (i < n1 && j < n2)
  {
    if (vec1[i] <= vec2[j])
    {
      vec[k] = vec1[i];
      i++;
    }
    else
    {
      vec[k] = vec2[j];
      j++;
    }
    k++;
  }

  while (i < n1)
  {
    vec[k] = vec1[i];
    i++;
    k++;
  }

  while (j < n2)
  {
    vec[k] = vec2[j];
    j++;
    k++;
  }
}

File: Go/merge_utils.go
package main

func Merge(arr []int, start, mid, end int) {
	leftSize := mid - start + 1
	rightSize := end - mid

	leftArr := make([]int, leftSize)
	for i := 0; i < leftSize; i++ {
		leftArr[i] = arr[start+i]
	}
	
	rightArr := make([]int, rightSize)
	for i := 0; i < rightSize; i++ {
		rightArr[i] = arr[mid+1+i]
	}

	i := 0
	j := 0
	k := start

	for i < leftSize && j < rightSize {
		if leftArr[i] < rightArr[j] {
			arr[k] = leftArr[i]
			i++
		} else {
			arr[k] = rightArr[j]
			j++
		}
		k++
	}

	for i < leftSize {
		arr[k] = leftArr[i]
		i++
		k++
	}

	for j < rightSize {
		arr[k] = rightArr[j]
		j++
		k++
	}
}